File: qwen_js/vectorization.js
import fs from 'fs';
import path from 'path';
import * as cheerio from 'cheerio';
import chromadb from 'chromadb';
import { PDFLoader } from "@langchain/community/document_loaders/fs/pdf";
import { text } from "stream/consumers"; //text brute
import { TextLoader } from "langchain/document_loaders/fs/text"; //text
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";


function add_context(list_file_path) {
    // --- CONFIGURATION & LECTURE JSON ---
    const configPath = path.resolve(proccess.cwd(), 'Config.json');
    const fileData = fs.readFileSync(configPath, 'utf-8');
    const config = JSON.parse(fileData);
    //Destination
    const DossierRagDB = config["VectDataBase"];
    console.log("1. Initialisation Chroma...");
    const embeddingsLocal = undefined;

    // Liste pour accumuler tous les morceaux de texte
    let tousLesChunks = [];
    const textSplitter = new RecursiveCharacterTextSplitter({
        chunkSize: 1000,
        chunkOverlap: 200
    });

    console.log("2. Lecture et découpage des fichiers PDF...");

// Vérification et création du dossier source (équivalent de os.makedirs)
    if (!fs.existsSync(DOSSIER_SRC)) {
        fs.mkdirSync(DOSSIER_SRC, { recursive: true });
        console.log(`Le dossier '${DOSSIER_SRC}' a été créé. Pensez à y mettre vos PDF !`);
    }

// Lecture du contenu du dossier (équivalent de os.listdir)
    const fichiers = fs.readdirSync(DOSSIER_SRC);

    for (const nomFichier of fichiers) {
        const cheminFichier = path.join(DOSSIER_SRC, nomFichier);
        const statut = fs.statSync(cheminFichier);

        // On ignore les dossiers (équivalent de os.path.isdir)
        if (statut.isDirectory()) {
            continue;
        }

        if (nomFichier.endsWith(".pdf")) {
            console.log(` -> Traitement de : ${nomFichier}`);
            try {
                const loader = new PDFLoader(cheminFichier);
                const docs = await loader.load();
                const chunks = await textSplitter.splitDocuments(docs);
                tousLesChunks.push(...chunks); // .push(...arr) équivaut au .extend() de Python
            } catch (e) {
                console.error(`    [Erreur] Impossible de lire le PDF ${nomFichier} :`, e.message);
            }
        }
        else if (nomFichier.endsWith(".htm") || nomFichier.endsWith(".html")) {
            console.log(` -> Traitement HTML : ${nomFichier}`);
            try {
                // 1. Lecture du fichier HTML (équivalent de open avec utf-8)
                const contenuHtml = fs.readFileSync(cheminFichier, 'utf-8');

                // 2. Utilisation de Cheerio (équivalent de BeautifulSoup)
                const $ = cheerio.load(contenuHtml);

                // Extraction du texte propre (en remplaçant les espaces blancs multiples)
                const textePropre = $('body').text().replace(/\s+/g, ' ').trim();

                // 3. Extraction du titre pour les métadonnées
                const titre = $('title').text() || nomFichier;

                // 4. Encapsulation dans un Document LangChain
                const documentNettoye = [
                    new Document({
                        pageContent: textePropre,
                        metadata: { source: nomFichier, title: titre }
                    })
                ];

                // 5. Découpage et ajout
                const chunks = await textSplitter.splitDocuments(documentNettoye);
                tousLesChunks.push(...chunks);
            } catch (e) {
                console.log(`    [Erreur] Impossible de lire le HTML ${nomFichier} : ${e.message}`);
            }
        }
        else {
            console.log(` -> Traitement (Mode Texte Brut) : ${nomFichier}`);
            try {
                // TextLoader en JS gère l'encodage textuel de base
                const loader = new TextLoader(cheminFichier);
                const docs = await loader.load();
                const chunks = await textSplitter.splitDocuments(docs);
                tousLesChunks.push(...chunks);
            } catch (e) {
                console.log(`    [Erreur] Impossible de lire le fichier ${nomFichier} en mode texte : ${e.message}`);
            }
        }
    }

// --- INITIALISATION BASE VECTORIELLE ---

    if (tousLesChunks.length === 0) {
        console.log("Attention : Aucun morceau de texte trouvé. Votre dossier source est-il vide ?");
    } else {
        console.log(`\n3. Création de la base vectorisée avec ${tousLesChunks.length} morceaux...`);

        // Extraction des données brutes
        const textes = tousLesChunks.map(doc => doc.pageContent);
        const metadatas = tousLesChunks.map(doc => doc.metadata);
        const ids = tousLesChunks.map((_, i) => `id_${i}`);

        try {
            // Utilisation du client persistant Chroma
            const client = new chromadb.ChromaClient({ path: DOSSIER_BASE_VECTORIELLE });

            const collection = await client.getOrCreateCollection({
                name: "ma_collection"
                // embeddingFunction: optionnel si tu veux laisser celle par défaut de Chroma
            });

            // Injection des données (Requiert de l'asynchrone en JS !)
            await collection.add({
                ids: ids,
                documents: textes,
                metadatas: metadatas
            });

            console.log(`\n[SUCCÈS] Votre dossier de base vectorisée a été créé : '${DOSSIER_BASE_VECTORIELLE}' !`);
        } catch (e) {
            console.error("Erreur lors de l'injection dans Chroma :", e);
        }
    }
}

File: Langchain_python/vectorization.py
import os
from email.headerregistry import UnstructuredHeader

from langchain_community.document_loaders import PyPDFLoader, BSHTMLLoader, TextLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_chroma import Chroma
# Changement d'import ici pour utiliser le moteur par défaut de Chroma
from chromadb.utils import embedding_functions

# --- CONFIGURATION ---
with open("Config.json", "r", encoding="utf-8") as f:
    config = json.load(f)

    # Extraction des variables de configuration
DOSSIER_SRC = config["SrcDataBase"]

# 2. C'est ICI que votre base de données vectorisée va être créée !
# Ce dossier va apparaître magiquement sur votre ordinateur.
DOSSIER_BASE_VECTORIELLE = config["VectDataBase"]

print("1. Initialisation du modèle d'embeddings interne de Chroma...")
# On utilise le modèle par défaut qui s'installe sans l'infrastructure HuggingFace
embeddings_local = embedding_functions.DefaultEmbeddingFunction()

# Liste pour accumuler tous les morceaux de texte
tous_les_chunks = []
text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)

print("2. Lecture et découpage des fichiers PDF...")
if not os.path.exists(DOSSIER_SRC):
    os.makedirs(DOSSIER_SRC)
    print(f"Le dossier '{DOSSIER_SRC}' a été créé. Pensez à y mettre vos PDF !")

for nom_fichier in os.listdir(DOSSIER_SRC):
    chemin_fichier = os.path.join(DOSSIER_SRC, nom_fichier)
    # On ignore les dossiers, on ne traite que les fichiers
    if os.path.isdir(chemin_fichier):
        continue

    if nom_fichier.endswith(".pdf"):
        print(f" -> Traitement de : {nom_fichier}")
        loader = PyPDFLoader(chemin_fichier)
        tous_les_chunks.extend(text_splitter.split_documents(loader.load()))

    elif nom_fichier.endswith(".htm") or nom_fichier.endswith(".html"):
        print(f" -> Traitement HTML : {nom_fichier}")
        try:
            from langchain_core.documents import Document
            from bs4 import BeautifulSoup
            # 1. On ouvre et on lit le fichier nous-mêmes en forçant l'UTF-8
            with open(chemin_fichier, "r", encoding="utf-8", errors="ignore") as f:
                contenu_html = f.read()
            # 2. On utilise BeautifulSoup pour nettoyer le HTML et extraire le texte propre
            soup = BeautifulSoup(contenu_html, "html.parser")
            texte_propre = soup.get_text(separator=" ")
            # 3. On extrait le titre de la page pour les métadonnées (comme le faisait BSHTMLLoader)
            titre = soup.title.string if soup.title else nom_fichier
            # 4. On encapsule le tout dans un objet Document standard de LangChain
            document_nettoye = [Document(page_content=texte_propre, metadata={"source": nom_fichier, "title": titre})]
            # 5. On découpe et on ajoute aux chunks
            tous_les_chunks.extend(text_splitter.split_documents(document_nettoye))
        except Exception as e:
            print(f"    [Erreur] Impossible de lire le HTML {nom_fichier} : {e}")
    else :
        print(f" -> Traitement (Mode Texte Brut) : {nom_fichier}")
        try:
            # TextLoader va forcer la lecture du fichier comme s'il s'agissait d'un .txt
            # autodetect_encoding=True évite les plantages avec les accents (UTF-8, Windows-1252...)
            loader = TextLoader(chemin_fichier, autodetect_encoding=True)
            tous_les_chunks.extend(text_splitter.split_documents(loader.load()))
        except Exception as e:
            print(f"    [Erreur] Impossible de lire le fichier {nom_fichier} en mode texte : {e}")


if len(tous_les_chunks) == 0:
    print("Attention : Aucun morceau de texte trouvé. Votre dossier 'RAG_src' est-il vide ?")
else:
    print(f"\n3. Création de la base vectorisée avec {len(tous_les_chunks)} morceaux...")

    # Extraction des textes et métadonnées bruts pour Chroma natif
    textes = [doc.page_content for doc in tous_les_chunks]
    metadatas = [doc.metadata for doc in tous_les_chunks]
    ids = [f"id_{i}" for i in range(len(tous_les_chunks))]

    # Utilisation directe du client persistant de Chroma
    import chromadb

    client = chromadb.PersistentClient(path=DOSSIER_BASE_VECTORIELLE)
    collection = client.get_or_create_collection(
        name="ma_collection",
        embedding_function=embeddings_local
    )

    # Injection des données
    collection.add(documents=textes, metadatas=metadatas, ids=ids)

    print(f"\n[SUCCÈS] Votre dossier de base vectorisée a été créé : '{DOSSIER_BASE_VECTORIELLE}' !")